app/notification_email.go

```diff
@@ -283,6 +287,36 @@ func getFormattedPostTime(user *model.User, post *model.Post, useMilitaryTime bo
 	}
 }
 
+func (a *App) generateHyperlinkForChannels(postMessage, teamName, teamURL string) string {
+	var publicChannels *model.ChannelList
+	if team, err := a.GetTeamByName(teamName); err != nil {
+		mlog.Error(fmt.Sprintf("Encountered error while looking up team by name for %s", teamName))
+		return postMessage
+	} else {
+		if publicChannels, err = a.GetPublicChannelsForTeam(team.Id, 0, 100); err != nil {
+			mlog.Error(fmt.Sprintf("Encountered error while looking up Public Channels for Team: %s", teamName))
+			return postMessage
+		}
+	}
+
+	r := regexp.MustCompile(`\~[\w-]+\b`)
+	matches := r.FindAllString(postMessage, -1)
+	var channelURL, channelHyperLink string
+	channelNameMap := make(map[string]struct{})
+	for _, channelName := range matches {
+		channelNameMap[channelName[1:]] = struct{}{}
+	}
+
+	for _, ch := range *publicChannels {
+		if _, ok := channelNameMap[ch.DisplayName]; ok {
+			channelURL = teamURL + "/channels/" + ch.Name
+			channelHyperLink = fmt.Sprintf("<a href='%s'>%s</a>", channelURL, "~"+ch.DisplayName)
+			postMessage = strings.Replace(postMessage, "~"+ch.DisplayName, channelHyperLink, -1)
```

Comment:
Simply doing a find-and-replace on the entire post is going to cause problems because it might find channel names in places where it shouldn't such as inside code blocks or as part of URLs. If you take a look at [`model.RewriteImageURLs`](https://github.com/mattermost/mattermost-server/blob/master/model/post.go#L466), you can see how it uses `markdown.Inspect` to parse through the post, find what it's looking for, and then build a new string with the text replaced. In your case, you'll end up looking for `*markdown.Text` nodes and doing a find-and-replace on them.

Reply:
Actually, I don't think this should matter. We don't render Markdown in emails anyway, so it's probably overkill to deal with this right now. I think it's fine to render channel links inside of Markdown for now